chain/chain/src/chain.rs

```diff
@@ -3222,34 +3295,150 @@ impl<'a> ChainUpdate<'a> {
-                    // only state roots are needed for updating split states
-                    self.chain_store_update.save_chunk_extra(
-                        &block_hash,
-                        &result.shard_uid,
-                        ChunkExtra::new_with_only_state_root(&result.new_root),
-                    );
-                    self.chain_store_update.save_trie_changes(result.trie_changes);
+                if is_new_chunk {
+                    // Split validator_proposals, gas_burnt, balance_burnt to each split shard
+                    // and store the chunk extra for split shards
+                    // Note that here we do not split outcomes by the new shard layout, we simply store
+                    // the outcome_root from the parent shard. This is because outcome proofs are
+                    // generated per shard using the old shard layout and stored in the database.
+                    // For these proofs to work, we must store the outcome root using per shard
+                    // using the old shard layout instead of the new shard layout
+                    let chunk_extra =
+                        self.chain_store_update.get_chunk_extra(block_hash, shard_uid)?;
+                    let next_epoch_shard_layout = {
+                        let epoch_id = self
+                            .runtime_adapter
+                            .get_next_epoch_id_from_prev_block(prev_block_hash)?;
+                        self.runtime_adapter.get_shard_layout(&epoch_id)?
+                    };
+
+                    let mut validator_proposals_by_shard: HashMap<_, Vec<_>> = HashMap::new();
+                    for validator_proposal in chunk_extra.validator_proposals() {
+                        let shard_id = account_id_to_shard_uid(
+                            &validator_proposal.account_id(),
+                            &next_epoch_shard_layout,
+                        );
+                        validator_proposals_by_shard
+                            .entry(shard_id)
+                            .or_default()
+                            .push(validator_proposal);
+                    }
+
+                    let num_split_shards = next_epoch_shard_layout
+                        .get_split_shards(shard_uid.shard_id())
+                        .unwrap_or_else(|| {
+                            panic!("invalid shard layout {:?}", next_epoch_shard_layout)
+                        })
+                        .len() as NumShards;
+                    let total_gas_used = chunk_extra.gas_used();
+                    let total_balance_burnt = chunk_extra.balance_burnt();
+                    let gas_res = total_gas_used % num_split_shards;
+                    let gas_split = total_gas_used / num_split_shards;
+                    let balance_res = (total_balance_burnt % num_split_shards as u128) as NumShards;
+                    let balance_split = total_balance_burnt / (num_split_shards as u128);
```

Comment:
is this actually the right way to split? What is the rationale behind it?

Reply:
I think gas_used and balance burnt are only used at the block level to calculate gas price, so how they are divided here actually does not matter as long as the sum remains the same

Reply:
hmm why wouldn't `validate_chunk_with_chunk_extra` break?

Reply:
Yeah, I'd probably expect to see something like 

```
fn split(total: u64, n_parts: usize) -> Vec<u64> {}
```

where the split is as even as possible (ie, the first rem items get +1)